### PR TITLE
OutputCSVFile for writing large CSV files.

### PR DESCRIPTION
**Pull request by voj**
We've started to create large rupture sets while exploring joint ruptures. This is mostly due to the large number of sections in our subduction faults.

This PR introduces a `CSVWriter` and a `CSVReader` that allow us to write the generated ruptures to file without requiring tens of gigabytes of heap. This is mainly achieved by not populating the `List<List<E>> values` property of `CSVFile`. The largest `indices.csv` file we have written with this so far was 4.9GB on disk.

### CODE REVIEW

**Comment by voj** [comment]
@kevinmilner hold on with reviewing this, I've got new ideas :-)

**Comment by kevinmilner** [comment]
> @kevinmilner hold on with reviewing this, I've got new ideas :-)

No problem @voj, I'm at a conference this week and won't get to it until next week anyway!

**Comment by voj** [comment]
All ready now!

File: src/main/java/org/opensha/commons/data/CSVReader.java
Comment (kevinmilner) [inline src/main/java/org/opensha/commons/data/CSVReader.java:1]: It looks like CSVReader reads the whole file into memory, did you consider making it just iterate over each line instead? That would require removing the getNumRows() method (and changing FaultSystemRupSet.loadRupSectsCSV(...) to use a while loop), but it should be even more memory efficient and maybe faster. It would also simplify things a bit as I'm finding LargeLineBuffer a bit hard to follow.

File: src/main/java/org/opensha/commons/data/CSVWriter.java
Comment (kevinmilner) [inline src/main/java/org/opensha/commons/data/CSVWriter.java:1]: It's not clear to me that having a `CSVWriter` interface (with it's only method being `writeToStream(OutputStream)` is a value add over just having it be a concrete class that does what `CSVWriterStream` currently does. The only method that I see that currently takes the interface rather than a specific implementation is the static `CSV_BackedModule.writeToArchive(...)` method that could just be duplicated which might simplify things? I don't feel strongly about this though, and could also be missing some current and/or future utility of setting it up this way. Mostly just putting this out here to see what you think.

File: src/main/java/org/opensha/commons/data/CSVReader.java
Comment (voj) [reply src/main/java/org/opensha/commons/data/CSVReader.java:1]: I had considered that but I wanted to respect the boundary that `CSV_BackedModule` enforces by being the only place that accesses the archive file. It opens and closes the stream and it' very clean. 
However, I have tried it out now and the code is much simpler to read. I've also re-written the reader to return a `Row` object to make it much easier to understand what's going on.

File: src/main/java/org/opensha/commons/data/CSVWriter.java
Comment (voj) [reply src/main/java/org/opensha/commons/data/CSVWriter.java:1]: After accepting that `CSVReader` could just wrap a zip stream directly, I also changed `CSVReader` to be the simplest `OutputStream` wrapper. Everything is much more straightforward now.

Some of the changes will be easier to see with whitespace hidden 
image

**Comment by voj** [comment]
@kevinmilner I've done a major rework after your comments

**Review by kevinmilner — COMMENTED** [review]
Looks great, much simpler. Thanks!

**Review by kevinmilner — APPROVED** [review]
